test/boost/reader_concurrency_semaphore_test.cc

```diff
@@ -1077,3 +1084,352 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_set_resources) {
+        auto s = tbl.schema();
+        auto& sst_man = tbl.get_sstables_manager();
+        auto& semaphore = db.get_reader_concurrency_semaphore();
+
+        dht::decorated_key dk = [&] () -> dht::decorated_key {
+            std::optional<dht::decorated_key> dk;
+            int32_t i = 0;
+            do {
+                dk = dht::decorate_key(*s, partition_key::from_single_value(*s, serialized(i++)));
+            } while (dht::shard_of(*s, dk->token()) != this_shard_id());
+            return std::move(*dk);
+        }();
+
+        mutation mut(s, dk);
+        mut.set_clustered_cell(clustering_key::from_single_value(*s, serialized(0)), to_bytes("value"), data_value(value), 0);
+
+        auto sstables_dir = tmpdir();
+
+        const auto target_num_sstables = 32;
+        const auto sstable_write_concurrency = 16;
+        const auto num_reads = 64;
+
+        auto num_sstables = 0;
+        parallel_for_each(boost::irange(0, sstable_write_concurrency), [&] (int i) {
+            return seastar::async([&, i] {
+                while (num_sstables != target_num_sstables) {
+                    ++num_sstables;
+                    auto sst = sst_man.make_sstable(s, sstables_dir.path().string(), sstables::generation_type{num_sstables}, sst_man.get_highest_supported_format(), sstables::sstable_format_types::big);
+                    auto writer_cfg = sst_man.configure_writer("test");
+                    sst->write_components(
+                        make_flat_mutation_reader_from_mutations_v2(s, semaphore.make_tracking_only_permit(s.get(), "test", db::no_timeout), mut, s->full_slice()),
+                        1,
+                        s,
+                        writer_cfg,
+                        encoding_stats{}).get();
+                    sst->load().get();
+                    tbl.add_sstable_and_update_cache(std::move(sst)).get();
+                }
+            });
+        }).get();
+
+        auto read_id = env.prepare("SELECT value FROM ks.tbl WHERE pk = ? AND ck = ?").get0();
+
+        // We first check that the test params are not too strict and a single
+        // read can finish successfully.
+        {
+            auto msg = env.execute_prepared(read_id, {cql3::raw_value::make_value(serialized(0)), cql3::raw_value::make_value(serialized(0))}).get();
+            assert_that(msg).is_rows().with_rows_ignore_order({ {serialized(value)} });
+        }
+
+        uint64_t successful_reads = 0;
+        uint64_t failed_reads = 0;
+
+        parallel_for_each(boost::irange(0, num_reads), [&] (int i) {
+            return env.execute_prepared(read_id, {cql3::raw_value::make_value(serialized(0)), cql3::raw_value::make_value(serialized(0))}).then_wrapped([&] (future<shared_ptr<cql_transport::messages::result_message>> fut) {
+                if (fut.failed()) {
+                    // We expect failed, OOM-killed reads here.
+                    // No way to verify why they failed so we swallow all failures.
```

Comment:
Can you check some statistics?

The reads could have failed due to real oom (bad_alloc), not oom-kill. Does the test not pass pre-patch?

Reply:
I do check that the OOM prevention mechanism was tripped, at least as many times as there are failed reads.
So we know all failed reads originate from intentional kills. This is not 100% accurate as a single read can trip the kill limit multiple times. Maybe we should have a per-permit counter. I'll think about it.

Reply:
It doesn't have to be accurate.

But, the test has to fail pre-series.

Reply:
(I mean crash due to OOM, not fail the counter checks)

Reply:
I couldn't achieve this in one test: have an OOM crash and have succeeding reads too.
So I split into two: an OOM crash test, checking that with protections enabled there is no crash anymore; and another test which checks that with the protections, all read failures can be attributed to the protections: read don't get as far to fail due to OOM.